### PR TITLE
Esy 0.3.x support

### LINKED ISSUE #166: Esy 0.3 support
Esy has changed the build output folder (it's under `_esy` now instead of `_build`). RLS needs to support that.

### PR DESCRIPTION
**Pull request by anmonteiro**
fixes #166

this works right now with hardcoded paths, but @andreypopp just mentioned to me that `esy command-env --json` and getting `cur__target_dir` will get the build directory where artifacts live. Probably worth changing to that.

Also cc @jordwalke, Andrey mentioned you were thinking about esy + editor integrations.

### CODE REVIEW

**Comment by andreypopp** [comment]
Yes, one can run `esy command-env --json` to obtain esy environment as JSON, then:

- `$cur__target_dir` points to a build directory
- `$PATH` can be used to resolve executables from the esy env
- `$OCAMLPATH` can be used to resolve ocamlfind packages from the esy env

I'm not sure what exactly RLS does so it might not need all of these above.

Note that inside `_esy` dir in a project there could be multiple subdirectories besides `default` for different sandboxes.

File: src/analyze/BuildSystem.re [hunk 1]
```diff
@@ -120,6 +124,17 @@ let detect = (rootPath, bsconfig) => {
   }) : Bsb(bsbVersion);
 };
 
+let getEsyCompiledBase = esyVersion =>
```
Comment (jordwalke) [inline src/analyze/BuildSystem.re]: Thanks, I was looking to make this change but now I don't have to!

I think you could just examine the directoreis to see which are available and give precedence to `_esy` over `_build`. That way there's less coupling on the esy binary being available.
The only requirement would be a directory structure be in place.

Comment (andreypopp) [reply src/analyze/BuildSystem.re]: Why do you think it's better than reading `$cur__target_dir` from the `esy command-env --json` output? I think this is the most robust way — it was available from the beginning — 0.0.x, 0.1.x, 0.2.x, 0.3.x had it unchanged.

File: src/analyze/BuildSystem.re [hunk 2]
```diff
@@ -14,8 +14,12 @@ type compilerVersion =
   | V406
   | V402;
 
+type packageManager =
+  /* TODO: this could have another constructor Opam */
+  | Esy(string);
```
Comment (jaredly) [inline src/analyze/BuildSystem.re]: sweet

File: src/analyze/BuildSystem.re [hunk 3]
```diff
@@ -120,6 +124,17 @@ let detect = (rootPath, bsconfig) => {
   }) : Bsb(bsbVersion);
 };
 
+let getEsyCompiledBase = esyVersion =>
+  switch (String.split_on_char('.', esyVersion)) {
+  | [_, minor, ..._] =>
+    if(int_of_string(minor) >= 3) {
+      "_esy" /+ "default" /+ "build"
+    } else {
+      "_build" /* TODO */
```
Comment (jaredly) [inline src/analyze/BuildSystem.re]: why the `TODO`?

**Comment by jaredly** [comment]
oh wait jk it was that esy didn't support 4.07 on windows. This can be merged now.

File: src/analyze/BuildSystem.re (hunk 3 above)
Comment (anmonteiro) [reply src/analyze/BuildSystem.re]: IDK! It was there before.

**Comment by jordwalke** [comment]
Thank you, everyone!

**Pull request merged**